base/sparse/sparsematrix.jl

```diff
@@ -1433,7 +1433,7 @@ function _broadcast_unary_nz2z_z2z_T{TvA,TiA,TvB}(f::Function, A::SparseMatrixCS
     return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
 end
 function _broadcast_unary_nz2z_z2z{Tv}(f::Function, A::SparseMatrixCSC{Tv})
-    _broadcast_unary_nz2z_z2z_T(f, A, Tv)
+    _broadcast_unary_nz2z_z2z_T(f, A, promote_op(f, Tv))
```

Comment:
Mystery solved. Specialization on the type of the function being broadcast is not occurring:
```julia
function structure{TvA,TiA}(A::SparseMatrixCSC{TvA,TiA})
    Bcolptr = Vector{TiA}(A.n + 1)
    Browval = Vector{TiA}(nnz(A))
    copy!(Bcolptr, 1, A.colptr, 1, A.n + 1)
    copy!(Browval, 1, A.rowval, 1, nnz(A))
    return Bcolptr, Browval
end
function oldbc{Tv}(f::Function, A::SparseMatrixCSC{Tv})
    oldbc_T(f, A, Base.promote_op(f, Tv))
end
function oldbc_T{TvA,TiA,TvB}(f::Function, A::SparseMatrixCSC{TvA,TiA}, ::Type{TvB})
    Bcolptr, Browval = structure(A)
    Bnzval = Vector{TvB}(nnz(A))
    @inbounds @simd for k in 1:nnz(A)
        Bnzval[k] = f(A.nzval[k])
    end
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end
function newbc{TvA,TiA}(f::Function, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr, Browval = structure(A)
    Bnzval = broadcast(f, view(A.nzval, 1:nnz(A)))
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end
function newbcnv{TvA,TiA}(f::Function, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr, Browval = structure(A)
    Bnzval = broadcast(f, A.nzval)
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end
function newbc_forcespec{TvA,TiA,FT<:Function}(f::FT, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr, Browval = structure(A)
    Bnzval = broadcast(f, view(A.nzval, 1:nnz(A)))
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end
function newbcnv_forcespec{TvA,TiA,FT<:Function}(f::FT, A::SparseMatrixCSC{TvA,TiA})
    Bcolptr, Browval = structure(A)
    Bnzval = broadcast(f, A.nzval)
    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
end

using BenchmarkTools

N = 10^2;
A = sprand(N, N, 0.1);
oldbcbench = @benchmark oldbc($abs, $A);
newbcbench = @benchmark newbc($abs, $A);
newbcnvbench = @benchmark newbcnv($abs, $A);
newbc_fsbench = @benchmark newbc_forcespec($abs, $A);
newbcnv_fsbench = @benchmark newbcnv_forcespec($abs, $A);

begin
    println("new with view versus old")
    println(ratio(median(newbcbench), median(oldbcbench)))
    println("new without view versus old")
    println(ratio(median(newbcnvbench), median(oldbcbench)))
    println("new with view and forced function-type specialization versus old")
    println(ratio(median(newbc_fsbench), median(oldbcbench)))
    println("new (without view) and forced function-type specialization versus old")
    println(ratio(median(newbcnv_fsbench), median(oldbcbench)))
end
```
yields
```julia
new with view versus old
BenchmarkTools.TrialRatio:
  time:             11.564562706270626
  gctime:           1.0
  memory:           1.0655270655270654
  allocs:           6.6
  time tolerance:   5.00%
  memory tolerance: 1.00%
new without view versus old
BenchmarkTools.TrialRatio:
  time:             11.552599009900991
  gctime:           1.0
  memory:           1.0626780626780628
  allocs:           6.4
  time tolerance:   5.00%
  memory tolerance: 1.00%
new with view and forced function-type specialization versus old
BenchmarkTools.TrialRatio:
  time:             1.5556930693069306
  gctime:           1.0
  memory:           1.0037986704653372
  allocs:           1.4
  time tolerance:   5.00%
  memory tolerance: 1.00%
new (without view) and forced function-type specialization versus old
BenchmarkTools.TrialRatio:
  time:             0.9927805280528053
  gctime:           1.0
  memory:           1.0009496676163343
  allocs:           1.2
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
So with forced specialization this approach performs well, though using a view incurs some overhead. Should forcing specialization be necessary? Best!